apps/emqx_message_transformation/src/emqx_message_transformation.erl

```diff
@@ -0,0 +1,820 @@
+    case emqx_utils_json:safe_encode(Payload) of
+        {ok, Bin} ->
+            {ok, Bin};
+        {error, Reason} ->
+            trace_failure(Transformation, "payload_encode_failed", #{
+                encoder => json,
+                reason => Reason
+            }),
+            error
+    end;
+encode(Payload, #{type := avro, schema := SerdeName}, Transformation) ->
+    try
+        {ok, emqx_schema_registry_serde:encode(SerdeName, Payload)}
+    catch
+        error:{serde_not_found, _} ->
+            trace_failure(Transformation, "payload_encode_schema_not_found", #{
+                encoder => avro,
+                schema_name => SerdeName
+            }),
+            error;
+        Class:Error:Stacktrace ->
+            trace_failure(Transformation, "payload_encode_schema_failure", #{
+                encoder => avro,
+                schema_name => SerdeName,
+                kind => Class,
+                reason => Error,
+                stacktrace => Stacktrace
+            }),
+            error
+    end;
+encode(
+    Payload, #{type := protobuf, schema := SerdeName, message_type := MessageType}, Transformation
+) ->
+    try
+        {ok, emqx_schema_registry_serde:encode(SerdeName, Payload, [MessageType])}
+    catch
+        error:{serde_not_found, _} ->
+            trace_failure(Transformation, "payload_encode_schema_not_found", #{
+                encoder => protobuf,
+                schema_name => SerdeName,
+                message_type => MessageType
+            }),
+            error;
+        Class:Error:Stacktrace ->
+            trace_failure(Transformation, "payload_encode_schema_failure", #{
+                encoder => protobuf,
+                schema_name => SerdeName,
+                message_type => MessageType,
+                kind => Class,
+                reason => Error,
+                stacktrace => Stacktrace
+            }),
+            error
+    end.
+
+trace_failure(#{log_failure := #{level := none}} = Transformation, _Msg, _Meta) ->
+    #{
+        name := _Name,
+        failure_action := _Action
+    } = Transformation,
+    ?tp(message_transformation_failed, _Meta#{log_level => none, name => _Name, message => _Msg}),
+    ok;
+trace_failure(#{log_failure := #{level := Level}} = Transformation, Msg, Meta0) ->
+    #{
+        name := Name,
+        failure_action := _Action
+    } = Transformation,
+    Meta = maps:merge(#{name => Name}, Meta0),
+    ?tp(message_transformation_failed, Meta#{
+        log_level => Level, name => Name, action => _Action, message => Msg
+    }),
+    ?TRACE(Level, ?TRACE_TAG, Msg, Meta).
+
+run_message_transformation_failed_hook(Message, Transformation) ->
+    #{name := Name} = Transformation,
+    TransformationContext = #{name => Name},
+    emqx_hooks:run('message.transformation_failed', [Message, TransformationContext]).
+
+is_payload_properly_encoded(#message{payload = Payload}) ->
+    try iolist_size(Payload) of
```

Comment:
Does it work for all the downstream messages processors if payload is NOT a `binary()`

Reply:
Theoretically, `#message.payload` is

- https://github.com/emqx/emqx/blob/bb9fb10ec54400076f4d43643a6722af9a2e6018/apps/emqx_utils/include/emqx_message.hrl#L36
- https://github.com/emqx/emqx/blob/bb9fb10ec54400076f4d43643a6722af9a2e6018/apps/emqx/src/emqx_types.erl#L240

What other downstream message processors are besides `emqx_frame`?  The latter seems to expect and return iolists:

- https://github.com/emqx/emqx/blob/bb9fb10ec54400076f4d43643a6722af9a2e6018/apps/emqx/src/emqx_frame.erl#L781